_posts/2025-05-16-last-ten-years.markdown

```diff
@@ -0,0 +1,137 @@
+I think our first public use of this (it was used a lot as an aside elsewhere) was our [2017 paper that extended type safety from Java code across the JNI boundary](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf).
+
+The `CCall` instruction at the time was actually a trap (with the expectation that a future version would move some or all of the trap handler into hardware).
+It branched to a handler, as did `CReturn`.
+The software stack in the paper used these to implement a trusted stack for cross-compartment calls.
+In CHERIoT, we use a very similar model but don't need a special instruction.
+We use the sealed entry (sentry) model, where one object type is reserved for capabilities that you can branch to but can't do anything else with.
+This was added a few years after the 2015 paper and meant that CHERIoT started from a model where the trap was not necessary.
+Using sentries means that our cross-compartment call can just use a jump to a sentry to invoke the switcher, rather than a trap.
+This is more scalable: a single system can provide multiple switchers like ours, but the 2015 paper required one per address space.
+In CHERIoT we extend sentries to control interrupt state and to differentiate between forward and backwards control-flow edges.
+
+If you look at the way that the ccall switcher works the 2015 paper, you'll see something that is quite recognisable from the CHERIoT compartment-switch routine.
+Both maintain a trusted stack and use it in similar ways.
+The paper expects each compartment to have its own stack (which made multithreading hard. Dapeng Gao [is doing exciting work here](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/202306-plarch-library-based-compartmentalisation.pdf)), whereas CHERIoT reuses and zeroes the stack.
+The CHERIoT approach is optimised for small systems.
+Modern microcontrollers are fast but have small amounts of memory.
+Stacks are small and so zeroing them is a low overhead, whereas requiring a lot more stack would put more pressure on the most scarce resource in the system.
+
+This paper also describes the pure-capability ABI.
+This lowers every pointer to a CHERI capability and was possible only because the offset had been added.
+Back then, I was calling this the 'sandbox ABI' because we couldn't use it for complete programs (Robert Watson convinced me that was a terrible name).
+It wasn't until a few years later that (led by Brooks Davis but with a lot of amazing work from others) we finally had a complete CHERI userspace.
+If you read only one CHERI paper, [the 2019 one describing that system](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201904-asplos-cheriabi.pdf) is probably the one you should read (this one won the ASPLOS Best Paper Award).
+
```

Comment:
FWIW, recently I’ve been giving the 2024 IEEE S&P article as the “read just one paper” because it summarises a lot of different aspects of the work, including highlighting both CHERIoT and Morello. It doesn’t go into any interesting technical depth, of course, and it also is focused primarily on memory safety without compartmentalisation. And I really quite like the CheriABI paper, which also doesn’t do compartmentalisation, so .. maybe you should stick with the one you’ve cited.